dbms/src/DataStreams/MultiversionSortedBlockInputStream.h

```diff
@@ -0,0 +1,197 @@
+#pragma once
+
+#include <common/logger_useful.h>
+
+#include <DataStreams/MergingSortedBlockInputStream.h>
+
+#include <deque>
+
+namespace DB
+{
+
+static const size_t MAX_ROWS_IN_MULTIVERSION_QUEUE = 8192;
+
+/* Deque with fixed memory size. Allows pushing gaps.
+ * frontGap() returns the number of gaps were inserted before front.
+ *
+ * Note: empty deque may have non-zero front gap.
+ */
+template <typename T>
+class FixedSizeDequeWithGaps
+{
+public:
+
+    struct ValueWithGap
+    {
+        size_t gap;
+        char value[sizeof(T)];
+    };
+
+    explicit FixedSizeDequeWithGaps(size_t size)
+    {
+        container.resize_fill(size + 1);
+    }
+
+    ~FixedSizeDequeWithGaps()
+    {
+        auto destruct_range = [this](size_t from, size_t to)
+        {
+            for (size_t i = from; i < to; ++i)
+                destructValue(i);
+        };
+
+        if (begin <= end)
+            destruct_range(begin, end);
+        else
+        {
+            destruct_range(0, end);
+            destruct_range(begin, container.size());
+        }
+    }
+
+    void pushBack(const T & value)
+    {
+        constructValue(end, value);
+        moveRight(end);
+        container[end].gap = 0;
+    }
+
+    void pushGap(size_t count) { container[end].gap += count; }
+
+    void popBack()
+    {
+        size_t curr_gap = container[end].gap;
```

Comment:
Is it ok that `container[end]` points to uninitialized data?

Reply:
Could you add a mention that gap of the current element is stored in the next element?

Reply:
Also, motivation of creation this structure would be helpful.

Reply:
`container[end]` is initialized by `container.resize_fill` at start and `container[end].gap = 0;` after push.